lib/ipni-client.js

```diff
@@ -0,0 +1,75 @@
+import { decodeBase64, decodeVarint } from '../vendor/deno-deps.js'
+
+/**
+ *
+ * @param {string} cid
+ * @returns {Promise<{
+ *  indexerResult: string;
+ *  provider?: { address: string; protocol: string };
+ * }>}
+ */
+export async function queryTheIndex (cid) {
+  const url = `https://cid.contact/cid/${encodeURIComponent(cid)}`
+
+  let providerResults
+  try {
+    const res = await fetch(url)
+    if (!res.ok) {
+      console.error('IPNI query failed, HTTP response: %s %s', res.status, await res.text())
+      return { indexerResult: `ERROR_${res.status}` }
+    }
+
+    const result = await res.json()
+    providerResults = result.MultihashResults.flatMap(r => r.ProviderResults)
+    console.log('IPNI returned %s provider results', providerResults.length)
+  } catch (err) {
+    console.error('IPNI query failed.', err)
+    return { indexerResult: 'ERROR_FETCH' }
+  }
+
+  let graphsyncProvider
+  for (const p of providerResults) {
+    // TODO: find only the contact advertised by the SP handling this deal
+    // See https://filecoinproject.slack.com/archives/C048DLT4LAF/p1699958601915269?thread_ts=1699956597.137929&cid=C048DLT4LAF
+    // bytes of CID of dag-cbor encoded DealProposal
+    // https://github.com/filecoin-project/boost/blob/main/indexprovider/wrapper.go#L168-L172
+    // https://github.com/filecoin-project/boost/blob/main/indexprovider/wrapper.go#L195
+
+    const [protocolCode] = decodeVarint(decodeBase64(p.Metadata))
+    const protocol = {
+      0x900: 'bitswap',
+      0x910: 'graphsync',
+      0x0920: 'http',
+      4128768: 'graphsync'
+    }[protocolCode]
+
+    const address = p.Provider.Addrs[0]
+    if (!address) continue
+
+    switch (protocol) {
+      case 'http':
+        return {
+          indexerResult: 'OK',
+          provider: { address, protocol }
+        }
+
+      case 'graphsync':
+        if (!graphsyncProvider) {
+          graphsyncProvider = {
+            address: `${address}/p2p/${p.Provider.ID}`,
+            protocol
+          }
+        }
+    }
+  }
+  if (graphsyncProvider) {
+    console.log('HTTP protocol is not advertised, falling back to Graphsync.')
+    return {
+      indexerResult: 'HTTP_NOT_ADVERTISED',
+      provider: graphsyncProvider
+    }
+  }
+
+  console.log('All advertisements are for unsupported protocols.')
+  return { indexerResult: 'NO_VALID_ADVERTISEMENT' }
```

Comment:
I am a bit concerned about the storage overhead this new field adds.

With 40 bytes per measurement and ~690k measurements per 20-minute round, we are adding ~27MB per round.

In the past, I was thinking about the following alternative: instead of adding a new field `indexerResult`, we can repurpose the existing field `statusCode` and invent our custom codes outside the range of valid HTTP response status codes.
- 900 for "ERROR_FETCH"
- 901 for "ipni returned 404"
- 902 for "ipni returned non-200 status"
- 903 for "NO_VALID_ADVERTISEMENT"
- and so on

The downside is that we cannot represent `HTTP_NOT_ADVERTISED` flag together with the actual status code returned by Lassie for the Graphsync retrieval. However, since we want to move to test exclusively HTTP retrievals, this should not be a big deal?

Having written that, a measurement has ~600 bytes now, so those extra ~40 bytes are ~7% of that. That seems reasonable to me and not worth worrying about.

@juliangruber WDYT?

Reply:
Can we store this as an enum in Postgres?

Reply:
> Can we store this as an enum in Postgres?

We could, but I am very reluctant to do so.

- Every time we add a new value in the spark checker code, we would have to add a DB schema migration script to spark-api.
- I am not concerned about extra space required in our DB because it's only temporary until the measurements are published.

I am more concerned about the amount of data we store in web3.storage.

Reply:
I see!

Unfortunately we can't tell about our quota there, since their backend times out before telling us our plan usage

Reply:
I agree that adding 7% is acceptable for the value we're getting.

An alternative would be to implement JS enums:

```js
const RESULTS = {
  'ERROR_FETCH': 0,
  // ...
}
```

Store an integer and map back to a string when downloading

Reply:
> Store an integer and map back to a string when downloading

That's an option too! It's a nice incremental & backwards-compatible change - we can easily support some measurements reporting `"ERROR_FETCH"` and other measurements reporting `0`. 

Let's defer this optimisation until we need it.